app/lib/analyzer/backend.dart

```diff
@@ -0,0 +1,158 @@
+      final PackageVersionAnalysis pva = list[0];
+      if (pva == null) return null;
+      analysis = pva.latestAnalysis;
+    }
+
+    // analysis was set
+    final Key analysisKey = versionKey.append(Analysis, id: analysis);
+    final List list = await db.lookup([analysisKey]);
+    return list[0];
+  }
+
+  /// Create a new [Analysis] object with the current timestamp.
+  Analysis prepareAnalysis(String package, String version) {
+    final DateTime now = new DateTime.now().toUtc();
+    final Key versionKey = db.emptyKey
+        .append(PackageAnalysis, id: package)
+        .append(PackageVersionAnalysis, id: version);
+    return new Analysis()
+      ..parentKey = versionKey
+      ..packageName = package
+      ..packageVersion = version
+      ..timestamp = now;
+  }
+
+  /// Stores the analysis, and either creates or updates its parent
+  /// [PackageAnalysis] and [PackageVersionAnalysis] rollups records.
+  Future storeAnalysis(Analysis analysis) async {
+    assert(analysis.id == null);
+    await db.commit(inserts: [analysis]);
+    assert(analysis.id != null);
+
+    // update package and version too
+    await db.withTransaction((Transaction tx) async {
+      final Key packageKey =
+          db.emptyKey.append(PackageAnalysis, id: analysis.packageName);
+      final Key packageVersionKey = packageKey.append(PackageVersionAnalysis,
+          id: analysis.packageVersion);
+      final List parents = await tx.lookup([packageKey, packageVersionKey]);
+      final PackageAnalysis package = parents[0];
+      final PackageVersionAnalysis version = parents[1];
+
+      final List<Model> inserts = [];
+      if (version == null) {
+        inserts.add(_createVersion(analysis));
+      } else {
+        if (_updateVersionLatest(version, analysis)) {
+          inserts.add(version);
+        }
+      }
+
+      // If there is no change in the version, there won't be a change in the
+      // package either.
+      //
+      // Note: [PackageAnalysis] doesn't store analyzerVersion, it relies on
+      // [PackageVersionAnalysis] to always store the latest.
+      if (inserts.isEmpty) {
+        await tx.rollback();
+        return;
+      }
+
+      if (package == null) {
+        inserts.add(_createPackage(analysis));
+      } else {
+        if (_updatePackageLatest(package, analysis)) {
+          inserts.add(package);
+        }
+      }
+
+      tx.queueMutations(inserts: inserts);
+      await tx.commit();
+    });
+  }
+
+  PackageAnalysis _createPackage(Analysis analysis) {
+    final package = new PackageAnalysis()
+      ..parentKey = db.emptyKey
+      ..id = analysis.packageName
+      ..latestVersionAnalysisKey = analysis.key;
+    return package;
+  }
+
+  bool _updatePackageLatest(PackageAnalysis package, Analysis analysis) {
+    if (isNewer(
+        analysis.semanticPackageVersion, package.latestSemanticVersion)) {
+      // current package version is older than latest known stable version
```

Comment:
This comment is not correct, since this is really comparing semantic versions (published package versions) not the `analysisVersion` (aka pana version).